ui/raidboss/emulator/ui/Tooltip.ts

```diff
@@ -1,44 +1,103 @@
+const hideEvents = [
+  'mouseleave',
+  'blur',
+] as const;
+
+const validDirections = [
+  'top',
+  'right',
+  'bottom',
+  'left',
+] as const;
+type ValidDirection = typeof validDirections[number];
+
+abstract class HTMLTemplateElement extends HTMLElement {
+  abstract get content(): DocumentFragment;
+}
+
+const showEvents = [
+  'mouseenter',
+  'focus',
+] as const;
+
+type TemplatesType = {[Property in ValidDirection]: HTMLTemplateElement};
+
+const toPx = (px: number): string => `${px}px`;
+
 export default class Tooltip {
-  constructor(target, direction, text, autoShow = true, autoHide = true) {
-    if (!Tooltip.validDirections.includes(direction))
-      throw new Error('Invalid direction for tooltip: ' + direction);
+  private offset = {
+    x: 0,
+    y: 0,
+  };
+  private target: HTMLElement;
+  private direction: ValidDirection;
+  private tooltip: HTMLElement;
+  private inner: HTMLElement;
+  private arrow: HTMLElement;
 
-    if (typeof target === 'string')
-      target = document.querySelector(target);
+  private static templates: TemplatesType;
+
+  constructor(
+      targetRef: string | HTMLElement,
+      direction: ValidDirection,
+      text: string,
+      autoShow = true,
+      autoHide = true) {
+    Tooltip.initializeTemplates();
+
+    let target: HTMLElement | null;
+
+    if (typeof targetRef === 'string')
+      target = document.querySelector<HTMLElement>(targetRef);
+    else
+      target = targetRef;
+
+    if (!target) {
+      const msg = 'Invalid selector or element passed to Tooltip';
+      console.error(msg);
+      throw new Error(msg);
+    }
 
-    this.offset = {
-      x: 0,
-      y: 0,
-    };
     this.target = target;
     this.direction = direction;
-    // Technically, templates should use querySelector, but this gets called so often
-    // on trigger-heavy encounters that we should hack its performance a bit
-    this.tooltip = document.getElementById(`${direction}TooltipTemplate`)
-      .content.firstElementChild.cloneNode(true);
+    this.tooltip = Tooltip.cloneTemplate(direction);
+    const innerElem = this.tooltip.querySelector<HTMLElement>('.tooltip-inner');
+    if (!innerElem) {
+      const msg = `Tooltip template direction ${direction} missing .tooltip-inner`;
+      console.error(msg);
+      throw new Error(msg);
+    }
+    this.inner = innerElem;
+    const arrowElem = this.tooltip.querySelector<HTMLElement>('.arrow');
+    if (!arrowElem) {
+      const msg = `Tooltip template direction ${direction} missing .arrow`;
+      console.error(msg);
+      throw new Error(msg);
```

Comment:
There is a lot of this pattern in this file.  One thing I've been doing elsewhere where something can't fail or will only fail if somebody breaks some major assumptions about the code like this is using `throw new UnreachableCode()`.  You could add a message parameter there if you wanted.  But, I've been using that for "this should ~never happen".

(I think this is different than the first `!target` check.  Having a caller pass in some name that may or may not exist seems like a failure case that should be handled [which is why I suggested not making the constructor handle it] but having a missing element in a template seems like something you should be able to just assume internally is true, if that make sense?)

Reply:
So to make sure I'm understanding correctly, `throw new UnreachableCode()` everywhere but the constructor's `!target` check?

Reply:
Yeah, that makes sense to me!  Feel free to disagree, but that's how I read this code.  Missing `target` feels like a plausible runtime error that should be handled but the others feel like "this should never happen".